File: algorithms/medium/maximum_length_of_a_concatenated_string_with_unique_characters.cpp
class Solution {
public:
    int maxLength(vector<string>& arr) {
      vector<bitset<26>> dp = { bitset<26>() };
      size_t ans = 0;
      for (string& s : arr) {
        bitset<26> a;
        for (char c : s) {
          a.set(c - 'a');
        }
        if (a.count() < s.size()) {
          continue;
        }
        for (int i = dp.size() - 1; i >= 0; --i) {
          if ((dp[i] & a).any()) {
            continue;
          }
          dp.push_back(dp[i] | a);
          ans = max(ans, (dp[i] | a).count());
        }
      }
      return (int)ans;
    }
};

File: algorithms/medium/maximum_length_of_a_concatenated_string_with_unique_characters.ex
defmodule Solution do
  @spec max_length(arr :: [String.t()]) :: integer
  def max_length(arr) do
    arr
    |> Enum.map(&String.graphemes/1)
    |> Enum.filter(fn w -> Enum.uniq(w) == w end)
    |> dfs(MapSet.new(), 0)
  end

  def dfs([], _, ans), do: ans

  def dfs([w | ws], set, ans) do
    w = MapSet.new(w)

    if MapSet.disjoint?(w, set) do
      new_set = MapSet.union(set, w)

      dfs(ws, new_set, max(ans, MapSet.size(new_set)))
      |> max(dfs(ws, set, ans))
    else
      dfs(ws, set, ans)
    end
  end
end
